File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.java
import java.util.*;
import java.io.*;


class Solution {
    int answer = 0;
    public int solution(int[] numbers, int target) {
        dfs(numbers, target, 0, -numbers[0]);
        dfs(numbers, target, 0 ,numbers[0]);
        return answer;
    }
    
    public void dfs(int[] numbers, int target, int depth, int sum){
        if(depth == numbers.length-1){
            if(sum == target){
                answer++;
            }
            return;
        }
        
        dfs(numbers, target, depth+1, sum+numbers[depth+1]);
        dfs(numbers, target, depth+1, sum-numbers[depth+1]);
    }
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cpp
#include <string>
#include <vector>

using namespace std;
vector<int> numbers;
int answer;
void dfs(int idx, int cur, int target) {
    if (idx == numbers.size()) {
        if (cur == target) {
            answer++;
        }
        return;
    }
    
    dfs(idx + 1, cur + numbers[idx], target);
    dfs(idx + 1, cur - numbers[idx], target);
}

int solution(vector<int> number, int target) {
    numbers = number;
    dfs(0, 0, target);
    return answer;
}